pkg/kubelet/kubelet.go

```diff
@@ -2014,6 +2014,10 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 		}
 	}
 
+	// compute and update the status in cache once the pods are no longer running
+	apiPodStatus = kl.generateAPIPodStatus(pod, podStatus, true)
```

Comment:
is there any reason why we need to call this both here as well as at the start of `SyncTerminatedPod`? I think conceptually it would make more sense to me to have this call just in `SyncTerminatedPod`.

i.e. at start of `SyncTerminatingPod`, containers are still running so we call `generateAPIPodStatus(..., false). Then when SyncTerminatingPod returns, we know all containers have been stopped. So at the next part of the state machine, `SyncTerminatedPod` we have the gurantee that no containers are running, so we call `generateAPIPodStatus(..., true). That way each sync method just has one call to `generateAPIPodStatus` and it uses only the gurantees of the previous function.

Reply:
Exiting SyncTerminatingPod frees other components to complete termination and cleanup - including eviction, which is free to delete containers once they reach completion (once SyncTerminatingPod exits).  In an earlier iteration we realized that was an issue, so this is bracketing a very very long synchronous call by establishing two status updates: one that we are terminating, and one that we have finished terminating, before other components in the kubelet are allowed to cleanup containers (so we get the terminal status).

Reply:
That makes sense thank you both for the explanation. Just for my understanding, what/where exactly are the containers removed between `SyncTerminatingPod` and `SyncTerminatedPod`? My understanding is that once `SyncTerminatingPod` exists, `IsPodKnownTerminated` is true, but there are [no callers](https://sourcegraph.com/search?q=context%3Aglobal+repo%3A%5Egithub%5C.com%2Fkubernetes%2Fkubernetes%24+file%3A%5Epkg%2Fkubelet+IsPodKnownTerminated&patternType=standard&sm=1&groupBy=path) of `IsPodKnownTerminated`.

Reply:
hm, possibly, in the one scenario that I have traced in more detail they were deleted by `podContainerDeletor`: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_container_deletor.go#L103.

But I think there exists an alternative path (I haven't cautght it though) by container GC in kuberuntime_gc.go

Reply:
Extended the comment to make the reason clearer.

Reply:
Thanks for updating the comment, lgtm. 

It makes sense that the status could be lost -- when we call `SyncTerminatedPod` we update the status from the runtime with `GetNewerThan` (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L1211). So that will mean that if the containers are deleted in between `SyncTerminatingPod` exiting and `SyncTerminatedPod`, then the exit codes and other info will be gone.

Reply:
When i merged the original pod worker we struggled with some races where status was lost, but we didn't have good enough testing to recreate it.  It was good that the tests in this PR exercised the logic enough that we could recreate it.

The GC can run at any time - the other scenario I saw was where https://github.com/kubernetes/kubernetes/blob/c015d6e434cc9b9b8882bda4c779708e8c2b151e/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L237 was getting called.  `ShouldPodRuntimeBeRemoved` returns true once syncTerminatingPod exits.  So the GC runs and the moment the